editor/plugins/curve_editor_plugin.cpp

```diff
@@ -311,236 +389,273 @@ void CurveEditor::on_preset_item_selected(int preset_id) {
-		}
+	int new_selected_index = selected_index;
+	// Reselect the old selected point if it's not the deleted one.
+	if (new_selected_index > p_index) {
+		new_selected_index -= 1;
+	} else if (new_selected_index == p_index) {
+		new_selected_index = -1;
 	}
 
-	return TANGENT_NONE;
-}
-
-void CurveEditor::add_point(Vector2 pos) {
-	ERR_FAIL_COND(_curve_ref.is_null());
-
 	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
 	undo_redo->create_action(TTR("Remove Curve Point"));
+	undo_redo->add_do_method(*curve, "remove_point", p_index);
+	undo_redo->add_do_method(this, "set_selected_index", new_selected_index);
+	undo_redo->add_undo_method(*curve, "add_point", old_pos, p.left_tangent, p.right_tangent, p.left_mode, p.right_mode);
+	undo_redo->add_undo_method(this, "set_selected_index", selected_index);
+	undo_redo->commit_action();
+}
 
-	Vector2 point_pos = get_world_pos(pos);
-	if (point_pos.y < 0.0) {
-		point_pos.y = 0.0;
-	} else if (point_pos.y > 1.0) {
-		point_pos.y = 1.0;
-	}
+void CurveEdit::set_point_position(int p_index, Vector2 p_pos) {
+	ERR_FAIL_COND(curve.is_null());
+	ERR_FAIL_INDEX_MSG(p_index, curve->get_point_count(), "Curve point is out of bounds.");
 
-	// Small trick to get the point index to feed the undo method
-	int i = _curve_ref->add_point(point_pos);
-	_curve_ref->remove_point(i);
+	// Use initial_grab_pos to determine things for the undo/redo.
+	if (initial_grab_pos == p_pos) {
+		return;
+	}
 
-	undo_redo->add_do_method(*_curve_ref, "add_point", point_pos);
-	undo_redo->add_undo_method(*_curve_ref, "remove_point", i);
+	// Pretend the point started from its old place.
+	curve->set_point_value(p_index, initial_grab_pos.y);
+	curve->set_point_offset(p_index, initial_grab_pos.x);
+	// Note: Changing the offset may modify the order.
+	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+	undo_redo->create_action(TTR("Modify Curve Point"));
+	undo_redo->add_do_method(*curve, "set_point_value", initial_grab_index, p_pos.y);
+	undo_redo->add_do_method(*curve, "set_point_offset", initial_grab_index, p_pos.x);
+	undo_redo->add_do_method(this, "set_selected_index", p_index);
+	undo_redo->add_undo_method(*curve, "set_point_value", p_index, initial_grab_pos.y);
+	undo_redo->add_undo_method(*curve, "set_point_offset", p_index, initial_grab_pos.x);
+	undo_redo->add_undo_method(this, "set_selected_index", initial_grab_index);
 	undo_redo->commit_action();
 }
 
-void CurveEditor::remove_point(int index) {
-	ERR_FAIL_COND(_curve_ref.is_null());
+void CurveEdit::set_point_tangents(int p_index, float p_left, float p_right) {
+	ERR_FAIL_COND(curve.is_null());
+	ERR_FAIL_INDEX_MSG(p_index, curve->get_point_count(), "Curve point is out of bounds.");
 
-	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
-	undo_redo->create_action(TTR("Remove Curve Point"));
-
-	Curve::Point p = _curve_ref->get_point(index);
+	if (initial_grab_left_tangent == p_left) {
```

Comment:
I only expect UndoRedo actions to trigger when the state of the system has changed. So I'm discarding these from UndoRedo.

Check out the 2D editor for example: Enable snapping, and move a texture around, returning it to where it was. This doesn't trigger an UndoRedo action.

Reply:
I don't think this comment was meant to go here, hehehe :D

Reply:
Nope, it's in the right place. If I change a value to... the old value. Then this check triggers and it's as if you never did it. No messages in the output etc. I think it's very handy.

Reply:
Oh, damn, I was still very sleepy and only read the first part of my comment so got confused 😴 

I see, so basically this just ensures that only the relevant UndoRedo get made, e.g. if we only moved the left tangent, no UndoRedo gets stored for the _right_ tangent. Is that right?

That being said, because of the issues with float, shouldn't `is_equal_approx()` be used? Or will this only happen if literally the tangent did not move at all, and therefore the value must be absolutely exactly the same, as opposed to "close enough"?